3p/recaptcha.js

```diff
@@ -124,28 +130,92 @@ function initializeIframeMessagingClient(window, grecaptcha, dataObject) {
  * action {string} - action to be dispatched with grecaptcha.execute
  * id {number} - id given to us by a counter in the recaptcha service
  *
+ * @param {Window} win
  * @param {*} grecaptcha
  * @param {Object} data
  */
-function actionTypeHandler(grecaptcha, data) {
-  // TODO: @torch2424: Verify message origin
-  // TODO: @torch2424: Verify sitekey is the same as original
-  const executePromise = grecaptcha.execute(data.sitekey, {
-    action: data.action,
-  });
+function actionTypeHandler(win, grecaptcha, data) {
+  doesOriginDomainMatchIframeSrc(win, data).then(() => {
+    // TODO: @torch2424: Verify sitekey is the same as original
+    const executePromise = grecaptcha.execute(data.sitekey, {
+      action: data.action,
+    });
 
-  // .then() promise pollyfilled by recaptcha api script
-  executePromise./*OK*/then(function(token) {
-    iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-token', dict({
-      'id': data.id,
-      'token': token,
-    }));
-  }, function(err) {
-    user().error(TAG, '%s', err.message);
-    iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-error', dict({
-      'id': data.id,
-      'error': err.message,
-    }));
+    // .then() promise pollyfilled by recaptcha api script
+    executePromise./*OK*/then(function(token) {
+      iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-token', dict({
+        'id': data.id,
+        'token': token,
+      }));
+    }, function(err) {
+      user().error(TAG, '%s', err.message);
+      iframeMessagingClient./*OK*/sendMessage('amp-recaptcha-error', dict({
+        'id': data.id,
+        'error': err.message,
+      }));
+    });
+  }).catch(error => {
+    dev().error(TAG, '%s', error.message);
   });
 }
 
+/**
+ * Function to verify our origin domain from the
+ * parent window.
+ * @param {Window} win
+ * @param {Object} data
+ * @return {!Promise}
+ */
+function doesOriginDomainMatchIframeSrc(win, data) {
+
+  if (!data.origin) {
+    return Promise.reject(
+        new Error('Could not retreive the origin domain')
+    );
+  }
+
+  const originLocation = parseUrlDeprecated(data.origin);
+
+  if (originLocation.hostname.includes('localhost') &&
+    win.location.hostname.includes('localhost')) {
+    return Promise.resolve();
+  }
+
+  if (isProxyOrigin(data.origin)) {
+    const curlsSubdomain = originLocation.hostname.split('.')[0];
```

Comment:
hmm, only comparing the domain prefix might lead to a security hole

a message from `victim-com.evil.com` would be accepted by `victim.com`

Reply:
Actually this is still good.

So as you know the two cases are: The bootstrap frame is hosted by a Cache AMP page. Or the bootstrap frames is hosted by a canonical AMP page.

The Cache AMP page cause is caught by `isProxyOrigin()`. Where the canonical case is covered by the statements after the `if()`.

In the canonical case, we are comparing the entire host, e.g `my.domain.example.com` to `my-domain-example-com`, which means you can't use a subdomain to get around this.

And what you are referring to is, what if they are in the cache case, and they try to compare `victim-com.evil.com` to `victim.com`. Well if we break the domain down into two parts, `[curlsdomain][cachedomain]`. The `evil.com` domain must be recognized as a `[cachedomain]`, and passes the cache regex within the `isProxyOrigin()` call that cache providers set up themselves.

When it fails, we will then drop to the canonical case, which will also reject. 😄 Feel free to discuss offline if you would like as well.

Reply:
I wrote a test for this, and thank you! In the canonical case, you were correct, I needed to check that the entire domain matches, thank you for mentioning this!

Reply:
Actually, in the canonical case, we were not checking the entire host like we should! But the idea still applies, fixing that now. Thanks for bringing this up! 👍

Reply:
@lannka Let's discuss this offline, because I am unsure if we want to force canonical AMP pages to still go through a cache for the recaptcha 3p frame.